### PR TITLE
feat: add tool input validation with graceful error handling

### LINKED ISSUE #2764: Agent doesn't check for tool param requirements
Currently:

   - Suppose a tool x requires param1 and param2. The agent doesn't check whether all params exist before calling the tool.
   - Throws an error after calling the tool, in runtime. 

Fix:

   - if a param doesn't exist, the agent should ask for more details so it can safely call the tool before failing in runtime

### PR DESCRIPTION
**Pull request by TylerBarnes**
This PR implements tool input validation that returns errors as results instead of throwing exceptions, allowing agents to understand validation failures and retry with corrected parameters.

The validation now uses Zod's safeParse method and returns structured error messages. When validation fails, tools return an error object with detailed field-level messages instead of throwing.

Before:
```typescript
// Would throw ZodError on invalid input
const result = await tool.execute({ email: 'invalid' });
```

After:
```typescript
// Returns error object on invalid input
const result = await tool.execute({ email: 'invalid' });
// result = {
//   error: true,
//   message: "Tool validation failed. Please fix the following errors and try again:\n- email: Invalid email format\n\nProvided arguments: { \"email\": \"invalid\" }",
//   validationErrors: { ... }
// }
```

For MCP servers, validation errors are returned with `isError: true` following the MCP spec, which allows LLMs to see the error and self-correct.

Fixes #2764

### CODE REVIEW

**Comment by TylerBarnes** [comment]
PR snapshot running here https://github.com/mastra-ai/mastra/actions/runs/16662126500

**Comment by TylerBarnes** [comment]
Claude says: Thanks for the review @greptileai\! I've addressed your feedback:

1. **Code duplication**: I extracted the validation logic into a shared `validateToolInput` utility function in `packages/core/src/tools/validation.ts`. Both the Tool class and CoreToolBuilder now use this shared function, eliminating the duplication and ensuring consistent behavior.

2. **Making execute private**: After discussing with the team, we decided to keep the current approach for now as it maintains the expected API.

3. **Integration tests**: I've added comprehensive integration tests in `packages/core/src/tools/validation.test.ts` that cover:
   - Required field validation
   - Type validation (number, boolean, string)
   - String constraints (email, min/max length, regex)
   - Complex types (arrays, objects, enums)
   - Data transformation after validation
   - Edge cases (no schema, empty context, passthrough)

The tests verify that validation errors are returned gracefully with detailed messages that help agents understand what went wrong.

All changes have been pushed in commits 42a20a5538 and a62e9d2db0.

- Claude

File: packages/core/src/tools/tool-builder/builder.ts [hunk 1]
```diff
@@ -147,16 +148,30 @@ export class CoreToolBuilder extends MastraBase {
       );
     };
 
-    return async (args: any, execOptions?: any) => {
+    return async (args: unknown, execOptions?: ToolExecutionOptions) => {
```
Comment (TylerBarnes) [inline packages/core/src/tools/tool-builder/builder.ts:151]: should args still be any?

File: packages/core/src/workflows/legacy/workflow-legacy.test.ts [hunk 2]
```diff
@@ -2314,7 +2314,11 @@ describe('LegacyWorkflow', async () => {
       });
 
       const workflow = new LegacyWorkflow({ name: 'test-workflow' });
-      workflow.step(step1).after(step1).step(randomTool).commit();
+      workflow
+        .step(step1)
+        .after(step1)
+        .step(randomTool, { variables: { name: { step: step1, path: 'name' } } })
```
Comment (TylerBarnes) [inline packages/core/src/workflows/legacy/workflow-legacy.test.ts:2320]: this should go back how it was

Comment (DanielSLew) [reply packages/core/src/workflows/legacy/workflow-legacy.test.ts:2320]: no, it causes a validation error if we don't pass on the correct args the tool needs. This test succeeded before because it just checks the tool.execute was called, but in this case it wasn't called because the error happens before the tool call rather than during

File: packages/core/src/tools/tool-builder/builder.ts (hunk 1 above)
Comment (DanielSLew) [reply packages/core/src/tools/tool-builder/builder.ts:151]: hmm not sure, I don't see any issues this causes, not sure why it was changed

File: packages/core/src/workflows/legacy/workflow-legacy.test.ts (hunk 2 above)
Comment (TylerBarnes) [reply packages/core/src/workflows/legacy/workflow-legacy.test.ts:2320]: ah got it, thanks!

**Review by DanielSLew — APPROVED** [review]
